File: core/apikeyworker/grpc.go
package apikeyworker

import (
	"context"
	"errors"

	"github.com/navidrome/navidrome/core/apikeyworker/gen"
	"github.com/navidrome/navidrome/core/rustworker"
	"google.golang.org/grpc"
)

var (
	apikeysGRPC = rustworker.NewManagedGRPC(rustworker.ManagedGRPCConfig{
		Name:   "apikeys",
		Listen: rustworker.DefaultListenAddr("navidrome-apikeys"),
		Resolve: func() (string, error) {
			return Resolve()
		},
		Health: func(ctx context.Context, conn *grpc.ClientConn) error {
			resp, err := gen.NewApiKeysClient(conn).Health(ctx, &gen.HealthRequest{})
			if err != nil {
				return err
			}
			if !resp.GetOk() {
				return errNoGRPC
			}
			return nil
		},
	})
	errNoGRPC = errors.New("apikeys gRPC worker unavailable")
)

// WarmGRPC starts the apikeys worker in the background for a warm first RPC.
func WarmGRPC() {
	apikeysGRPC.Warm()
}

func callAPIKeys[T any](ctx context.Context, fn func(context.Context, *grpc.ClientConn) (T, error)) (T, error) {
	result, err := rustworker.CallGRPC(apikeysGRPC, ctx, fn)
	if errors.Is(err, rustworker.ErrWorkerUnavailable) {
		var zero T
		return zero, errNoGRPC
	}
	return result, err
}

func generateGRPC(ctx context.Context, pepper string) (generateResult, error) {
	return callAPIKeys(ctx, func(ctx context.Context, conn *grpc.ClientConn) (generateResult, error) {
		cli := gen.NewApiKeysClient(conn)
		resp, err := cli.Generate(ctx, &gen.GenerateRequest{Pepper: pepper})
		if err != nil {
			return generateResult{}, err
		}
		if !resp.GetOk() {
			return generateResult{}, errors.New(nonEmpty(resp.GetError(), "Rust apikeys Generate failed"))
		}
		return generateResult{Token: resp.GetToken(), LookupPrefix: resp.GetLookupPrefix(), Hash: resp.GetHash()}, nil
	})
}

func hashGRPC(ctx context.Context, token, pepper string) (string, string, error) {
	type hashResult struct {
		prefix string
		hash   string
	}
	result, err := callAPIKeys(ctx, func(ctx context.Context, conn *grpc.ClientConn) (hashResult, error) {
		cli := gen.NewApiKeysClient(conn)
		resp, err := cli.Hash(ctx, &gen.HashRequest{Token: token, Pepper: pepper})
		if err != nil {
			return hashResult{}, err
		}
		if !resp.GetOk() {
			return hashResult{}, errors.New(nonEmpty(resp.GetError(), "Rust apikeys Hash failed"))
		}
		return hashResult{prefix: resp.GetLookupPrefix(), hash: resp.GetHash()}, nil
	})
	if err != nil {
		return "", "", err
	}
	return result.prefix, result.hash, nil
}

func verifyGRPC(ctx context.Context, token, hash, pepper string) (bool, error) {
	return callAPIKeys(ctx, func(ctx context.Context, conn *grpc.ClientConn) (bool, error) {
		cli := gen.NewApiKeysClient(conn)
		resp, err := cli.Verify(ctx, &gen.VerifyRequest{Token: token, Hash: hash, Pepper: pepper})
		if err != nil {
			return false, err
		}
		if !resp.GetOk() {
			return false, errors.New(nonEmpty(resp.GetError(), "Rust apikeys Verify failed"))
		}
		return resp.GetValid(), nil
	})
}

func nonEmpty(v, fallback string) string {
	if v == "" {
		return fallback
	}
	return v
}

File: rust/apikeys/src/grpc.rs
use tonic::{Request, Response, Status};

use crate::proto::api_keys_server::ApiKeys;
use crate::proto::{
    GenerateRequest, GenerateResponse, HashRequest, HashResponse, HealthRequest, HealthResponse,
    VerifyRequest, VerifyResponse,
};
use crate::{generate, hash_token, verify_token};

#[derive(Default, Clone)]
pub struct ApiKeysService;

#[tonic::async_trait]
impl ApiKeys for ApiKeysService {
    async fn generate(
        &self,
        request: Request<GenerateRequest>,
    ) -> Result<Response<GenerateResponse>, Status> {
        let pepper = request.into_inner().pepper;
        match generate(&pepper) {
            Ok((token, lookup_prefix, hash)) => Ok(Response::new(GenerateResponse {
                ok: true,
                token,
                lookup_prefix,
                hash,
                error: String::new(),
            })),
            Err(error) => Ok(Response::new(GenerateResponse {
                ok: false,
                error: error.to_string(),
                ..Default::default()
            })),
        }
    }

    async fn hash(
        &self,
        request: Request<HashRequest>,
    ) -> Result<Response<HashResponse>, Status> {
        let req = request.into_inner();
        match hash_token(&req.token, &req.pepper) {
            Ok(hash) => Ok(Response::new(HashResponse {
                ok: true,
                lookup_prefix: crate::lookup_prefix(&req.token),
                hash,
                error: String::new(),
            })),
            Err(error) => Ok(Response::new(HashResponse {
                ok: false,
                error: error.to_string(),
                ..Default::default()
            })),
        }
    }

    async fn verify(
        &self,
        request: Request<VerifyRequest>,
    ) -> Result<Response<VerifyResponse>, Status> {
        let req = request.into_inner();
        match verify_token(&req.token, &req.hash, &req.pepper) {
            Ok(valid) => Ok(Response::new(VerifyResponse {
                ok: true,
                valid,
                error: String::new(),
            })),
            Err(error) => Ok(Response::new(VerifyResponse {
                ok: false,
                error: error.to_string(),
                ..Default::default()
            })),
        }
    }

    async fn health(
        &self,
        _request: Request<HealthRequest>,
    ) -> Result<Response<HealthResponse>, Status> {
        Ok(Response::new(HealthResponse { ok: true }))
    }
}
